clippy_lints/src/methods/useless_asref.rs

```diff
@@ -111,6 +111,23 @@ fn is_calling_clone(cx: &LateContext<'_>, arg: &hir::Expr<'_>) -> bool {
         hir::ExprKind::Closure(&hir::Closure { body, .. }) => {
             // If it's a closure, we need to check what is called.
             let closure_body = cx.tcx.hir().body(body);
+
+            // |x| ...
+            //  ^
+            let [
+                hir::Param {
+                    pat:
+                        hir::Pat {
+                            kind: hir::PatKind::Binding(_, local_id, ..),
```

Comment:
You also need to check for `Ref`.

Reply:
I assume you mean `|&x|` patterns (`PatKind::Ref`)? Do you mean that we should also lint on `.as_ref().map(|&x| x.clone())`? Or do you mean that we shouldn't lint when `ref` is on the binding?

Reply:
I'm not sure exactly what this `Ref` pattern is checking exactly. I looked at `map_clone` (which lint for something similar) to check how they did and they handle both so I think you should as well here.

Reply:
Hmm, looking more into this, (correct me if I'm wrong) it looks like map_clone *used to* lint on `.as_ref().map(|&x| x)` because it's the same as `.copied()`, but as of recently doesn't anymore, because of a `should_run_lint` check that lets useless_asref check it.
But useless_asref only seems to handle `.clone()` calls, so there's nothing to catch that case anymore. 

I suppose we could extend the logic in `should_run_lint` to only return false when clone is called in the closure, so that map_clone can continue linting on `.map(|x| *x)` and `.map(|&x| x)`. We won't need a `Ref` check here, then (because it's already handled elsewhere).

This is getting pretty close to duplicating the logic from useless_asref inside should_run_lint though. I'm wondering if we could just move the clone checking that's currently in useless_asref to map_clone so they don't need this sort of coordination?

Reply:
Oh but you still can trick the lint:

```rust
let x = Some(String::new());
let x = x.as_ref();
x.map(|&x| x.clone()); // or whatever
```

If this is duplicating the logic from `map_clone`, then might be worth having some code in common instead. :)

Reply:
Ah yea I see what you mean now with `Ref`, I'll make it lint this too.
I think fixing that other false negative mentioned is a little more involved so it might be better done in a separate PR?

Reply:
As long as `Ref` is handled in this PR, I think it's fine.

Reply:
Fixed in https://github.com/rust-lang/rust-clippy/pull/12136/commits/be5707ce1b608f607ca9327537076dc95869347b . Is that what you had in mind? (This also moved that let else from before to an if let guard because I think it looks nicer and less indentation, shouldn't change anything functionally)

Reply:
opened a new issue for that one: #12142, so this can't be forgotten :)

Reply:
Unless I missed something, you still only handle `Binding` and still not `Ref` (`Ref` is used [here](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/methods/map_clone.rs#L53) in `map_clone`).

Reply:
https://github.com/rust-lang/rust-clippy/blob/be5707ce1b608f607ca9327537076dc95869347b/clippy_lints/src/methods/useless_asref.rs#L117

This removes any ref patterns, so `|&&x|`, `|&x|`, `|x|` will all be handled the same way

Reply:
Oh I see. Nicely done!